apps/omg_api/lib/state/core.ex

```diff
@@ -168,98 +163,111 @@ defmodule OMG.API.State.Core do
+    do: get_input_utxos(utxos, acc, inputs)
+
+  defp get_input_utxos(utxos, acc, [%{blknum: blknum, txindex: txindex, oindex: oindex} | inputs]) do
+    position = Utxo.position(blknum, txindex, oindex)
 
-  defp get_utxo(utxos, position) do
     case Map.get(utxos, position) do
       nil -> {:error, :utxo_not_found}
-      found -> {:ok, found}
+      found -> get_input_utxos(utxos, [found | acc], inputs)
     end
   end
 
-  defp is_spender?(owner, spender) do
-    if owner == spender, do: :ok, else: {:error, :incorrect_spender}
+  defp get_amounts_by_currency(utxos) do
+    utxos
+    |> Enum.group_by(fn %{currency: currency} -> currency end, fn %{amount: amount} -> amount end)
+    |> Enum.map(fn {currency, amounts} -> {currency, Enum.sum(amounts)} end)
+    |> Map.new()
+  end
+
+  # fee is implicit - it's the difference between funds owned and spend
+  defp amounts_add_up?(input_amounts, output_amounts, fees) do
+    with :ok <- same_currencies?(input_amounts, output_amounts) do
+      amounts_add_up =
+        for {input_currency, input_amount} <- Map.to_list(input_amounts) do
+          fee = Map.get(fees, input_currency)
+          output_amount = Map.get(output_amounts, input_currency)
+          input_amount >= fee + output_amount
+        end
+        |> Enum.all?()
+
+      if amounts_add_up, do: :ok, else: {:error, :amounts_do_not_add_up}
+    end
   end
 
-  defp same_currency?(utxo_currency, spent_currency) do
-    if utxo_currency == spent_currency, do: :ok, else: {:error, :incorrect_currency}
+  defp same_currencies?(input_amounts, output_amounts) do
+    if Map.keys(input_amounts) == Map.keys(output_amounts) do
+      :ok
+    else
+      {:error, :input_and_output_currencies_do_not_match}
+    end
   end
 
-  # fee is implicit - it's the difference between funds owned and spend
-  defp amounts_add_up?(has, spends) do
-    if has >= spends, do: :ok, else: {:error, :amounts_dont_add_up}
+  defp add_pending_tx(%Core{pending_txs: pending_txs, tx_index: tx_index} = state, new_tx) do
+    %Core{
+      state
+      | tx_index: tx_index + 1,
+        pending_txs: [new_tx | pending_txs]
+    }
   end
 
   defp apply_spend(
          %Core{height: height, tx_index: tx_index, utxos: utxos} = state,
-         %Transaction{
-           blknum1: blknum1,
-           txindex1: txindex1,
-           oindex1: oindex1,
-           blknum2: blknum2,
-           txindex2: txindex2,
-           oindex2: oindex2
-         } = tx
+         %Transaction{inputs: inputs} = tx
        ) do
     new_utxos_map =
       tx
       |> non_zero_utxos_from(height, tx_index)
       |> Map.new()
 
-    %Core{
-      state
-      | utxos:
-          utxos
-          |> Map.delete(Utxo.position(blknum1, txindex1, oindex1))
-          |> Map.delete(Utxo.position(blknum2, txindex2, oindex2))
-          |> Map.merge(new_utxos_map)
-    }
+    utxos =
+      inputs
+      |> Enum.reduce(utxos, fn %{blknum: blknum, txindex: txindex, oindex: oindex}, utxos ->
```

Comment:
as soon as you get rid of the map-representation of utxo-pos you can do `Map.drop`